### PR TITLE
 SETI-1210: Zuul can mix PRs from different projects

### PR DESCRIPTION
**Pull request by janpriessnitz**
Zuul only checks changes by PR number when deleting the old ones.
Check project name too in isUpdateOf()

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/70/ : SUCCESS in 12s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/82/ : FAILURE in 9m 40s

**Comment by janpriessnitz** [comment]
ok to test

**Comment by janpriessnitz** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/73/ : SUCCESS in 9s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/85/ : SUCCESS in 9m 49s

File: zuul/scheduler.py [hunk 1]
```diff
@@ -1395,19 +1395,20 @@ def getItemForChange(self, change):
...
+                result.append(item)
+        return result
 
     def removeOldVersionsOfChange(self, change):
         if not self.pipeline.dequeue_on_new_patchset:
             return
-        old_item = self.findOldVersionOfChangeAlreadyInQueue(change)
-        if old_item:
+        old_items = self.findOldVersionsOfChangeAlreadyInQueue(change)
+        for old_item in old_items:
```
Comment (hrubi) [inline zuul/scheduler.py]: I'm curious how this can happen that there are several change versions in the queue for a single change?

File: tests/test_github.py [hunk 2]
```diff
@@ -542,3 +542,27 @@ def test_custom_github_status_name(self):
...
 
         self.assertIn('abcd', A.statuses)
+
+    def test_multiple_projects(self):
+        "Test that Zuul does not delete old versions of PRs of other projects"
+        self.worker.hold_jobs_in_build = True
+
+        A = self.fake_github.openFakePullRequest('org/project', 'master', 'A')
+        self.fake_github.emitEvent(A.getPullRequestOpenedEvent())
+        time.sleep(2)
```
Comment (hrubi) [inline tests/test_github.py]: Why the sleep? It slows down tests, so it's generally not good if you can avoid it.

File: tests/test_github.py [hunk 3]
```diff
@@ -542,3 +542,27 @@ def test_custom_github_status_name(self):
         self.waitUntilSettled()
 
         self.assertIn('abcd', A.statuses)
+
+    def test_multiple_projects(self):
```
Comment (hrubi) [inline tests/test_github.py]: Awesome, thanks for covering via test!

File: zuul/scheduler.py (hunk 1 above)
Comment (janpriessnitz) [reply zuul/scheduler.py]: I think generally it cannot happen. However, I didn't think long about the corner cases. The main reason was that the function is named removeOldVersions...., so it can be pretty confusing when it deletes only one. Also, I think it's more robust, although it can be a little bit slower.

File: tests/test_github.py (hunk 2 above)
Comment (janpriessnitz) [reply tests/test_github.py]: The reason is Zuul considers a newer PR an update of an older PR only if it's got newer timestamp, but the timestamp is formatted to whole seconds, so it does not delete the old PR unless the new one is created at least a second later.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/78/ : SUCCESS in 46s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/90/ : FAILURE in 10m 06s

File: tests/test_github.py (hunk 2 above)
Comment (hrubi) [reply tests/test_github.py]: Right, I didn't think about that. So is 1 second enough? :)

**Comment by janpriessnitz** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/79/ : SUCCESS in 8s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/91/ : FAILURE in 7m 25s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/80/ : SUCCESS in 8s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/92/ : SUCCESS in 11m 38s

File: zuul/scheduler.py (hunk 1 above)
Comment (hrubi) [reply zuul/scheduler.py]: I think that it can't happen whatsoever. When a new item is about to enter the queue, this purging is done. So there's no way how two or more items for a single change would be already in queue.

Comment (janpriessnitz) [reply zuul/scheduler.py]: Well, there's a chance that two events triggering check pipeline occur in the same second. Then the newer one does not remove the older one, because the timestamps are equal.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/89/ : SUCCESS in 46s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/102/ : FAILURE in 11m 27s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/91/ : SUCCESS in 20s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/104/ : FAILURE in 9m 20s

File: zuul/scheduler.py (hunk 1 above)
Comment (hrubi) [reply zuul/scheduler.py]: Well, that actually is a bug on its own, and your solution won't solve such situation, when two events come at the same time and another event won't come at all. Could you think about solving it somehow differently?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/93/ : SUCCESS in 8s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/107/ : SUCCESS in 7m 58s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/94/ : SUCCESS in 7s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/109/ : SUCCESS in 9m 25s

File: tests/base.py [hunk 4]
```diff
@@ -701,18 +701,20 @@ def __init__(self, connection_name, connection_config, upstream_root=None):
...
         self.upstream_root = upstream_root
         self.merge_failure = False
         self.merge_not_allowed_count = 0
         self.files = {}
 
     def openFakePullRequest(self, project, branch, subject, files=[]):
-        self.pr_number += 1
+        if project not in self.pr_number:
+            self.pr_number[project] = 0
+        self.pr_number[project] += 1
```
Comment (hrubi) [inline tests/base.py]: What about this:
```
self.pr_number[project] = self.pr_number.setdefault(project, 0) + 1
```

File: tests/test_github.py [hunk 5]
```diff
@@ -551,3 +551,28 @@ def test_base_changed_event(self):
...
         self.assertEqual(A.statuses['check']['state'], 'success')
+
+    def test_multiple_projects(self):
+        "Test that Zuul does not delete old versions of PRs of other projects"
+        self.worker.hold_jobs_in_build = True
+
+        A = self.fake_github.openFakePullRequest('org/project', 'master', 'A')
+        self.fake_github.emitEvent(A.getPullRequestOpenedEvent())
+        # New PR has to be at least a second newer to be considered as an
+        # update of an old one.
```
Comment (hrubi) [inline tests/test_github.py]: No longer true. Remove.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/96/ : SUCCESS in 9s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/111/ : SUCCESS in 6m 52s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/zuul-merge-compliance-check/14/ : SUCCESS in 31s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/112/ : SUCCESS in 13m 20s
